### ISSUE TITLE
Automated Backups

### ISSUE THREAD

**Issue by LiamMorrow** [LiamMorrow/LiftLog#243]
Right now backups are a very manual process.

It would be good to have some sort of automated backup system.

It should:
- Be able to backup to an arbitrary directory outside of LiftLog's normal storage area
- Be able to run on a schedule (although saving to files is pretty quick, it might just be better to have it save everytime a session is saved)
- Store multiple backup files - timestamped - with a deletion policy?

**Comment by LiamMorrow** [comment]
Local storage backups seem like it would have to be an android only feature, which I don't love.  Not a blocker by any means. Just wondering if there's something we could do there.

In the past I have used apps which automatically backup to icloud and google storage.

I wonder if this could be done in such a way where there is a backup "silo" where a silo is something which depends on the platform:
- Android - Google Drive (maybe local storage, but it seems like it's getting harder and harder)
- iOS - iCloud.


@bschwehn would this help you somwhat?  You mentioned potentially running a feed server of your own which makes me think you'd much rather a fully self hosted solution.  Perhaps (and no guarantees), FTP?  Or even just a webhook.

**Comment by LiamMorrow** [comment]
![image](https://github.com/user-attachments/assets/65b3fedc-6a08-46f3-89b1-d99bfcd104de)
Just spitballing

**Comment by bschwehn** [comment]
> Local storage backups seem like it would have to be an android only feature, which I don't love.
> (from 244)
> especially as this is a feature that I likely wouldn't use.

Then I think the most important aspect is that you keep enjoying to work on this application and that it not starts feeling like work were you have to support other people's requirements (without even getting paid for it) :)

> makes me think you'd much rather a fully self hosted solution

Right, I prefer minimizing dependencies on third party dependencies and "always connected" solutions when I reasonably can. And I like to tinker with my own pipelines (like automatic export - syncthing to own server - convert and archive).

But, realistically, only a small fraction of your users will care a lot about automated backups at all. And an even smaller fraction of those users would care about maintaining their own infrastructure to enable this. Maybe it is even just one potential user: I.

Having said that, that webhook backup looks great to me, I'd use it :)
I'd also use a Google Drive backup, and the fractions of people that want automated backups and use Android without google is surely very small.

Now, I don't really have a good idea about the amount of implementation and maintenance efforts that would be required to add backup to Google Drive for Android and iCloud for iOS. At the least, it will require testing this in both platforms rather than being able to test it in the web application, so much more hassle. It would be more useful to more people compared to webhook, however.

If you think the webhook backup can be implemented quite easily, maybe it's something to start with and then wait until you have the feeling the automated backup is something that more users would like to use?

Again, I think I might be the only user right now that would want to use this, so thank you for still entertaining this idea! And I totally understand if this is not something you feel is worth to add right now.

**Comment by LiamMorrow** [comment]
> Then I think the most important aspect is that you keep enjoying to work on this application and that it not starts feeling like work were you have to support other people's requirements (without even getting paid for it) :)

Oh don't stress about that! Having people interested in using something I made is greatly motivating.  My main point is I'd like to get it right rather than just guessing what you'd like. I myself would actually love an automated backup, having lost data due to me developing the app on my daily driver 😬 

> Having said that, that webhook backup looks great to me, I'd use it :)

Looking at it, the major issue I see is that a webhook would require HTTPS in later versions of android. Not huge, I think at this stage getting a TLS cert from letsencrypt is pretty easy, but still something to note.

> Now, I don't really have a good idea about the amount of implementation and maintenance efforts that would be required to add backup to Google Drive for Android and iCloud for iOS. At the least, it will require testing this in both platforms rather than being able to test it in the web application, so much more hassle. It would be more useful to more people compared to webhook, however.

Yeah I think that it should be *fine* my main goal is that whatever solution that is done can be adapted to each platform (e.g. local backup wouldn't work on  ios, so is unattractive to me, but cloud backup works on ios and and android so is good.).  Web is less of an issue , it's kind of there for development purposes and a demo.  I'll tinker with the google drive API and see what I can do.  My fear is as with anything Apple development, iCloud will be a nightmare to authenticate with.

**TLDR**
I'd like to do the cloud backups as it would have a wider reach, but if it's too complex, the webhook is a good standin which would work on everything

**Comment by LiamMorrow** [comment]
@bschwehn I've merged #260 which implements the webhook variant of backups.  Have a read if you like.  I've provided a simple server `LiftLog.BackupServer` which can accept these backup requests and persist them to disk.

The major thing I see as a potential blocker is that the endpoint is required to be HTTPS.  Android (at least, I didn't check iOS) does not support unencrypted connections without a whitelist in the manifest.  I might be wrong here, and if this doesn't work for you and you find a workaround I'll be happy to try some things.

**Comment by LiamMorrow** [comment]
This has been released in https://github.com/LiamMorrow/LiftLog/releases/tag/2.17.2 🎉 

It's a little noisy, I might remove the toast on success.

**Comment by bschwehn** [comment]
Thanks @LiamMorrow, that looks great. 

I only just now got a chance to try it and it fails with [the connection failure error](https://github.com/LiamMorrow/LiftLog/blob/main/LiftLog.Ui/Store/Settings/SettingsEffects.cs#L257).

I wanted to use a server in my local network and use a self-signed cert and have added the CA to the the trusted CAs in android (no TLS error in Chrome in Android, so this part worked), but I think it still fails because it may need this setting in network_security_config in order for LiftLog to accept user CAs:

~~~xml
   <base-config>
        <trust-anchors>
        <!-- Trust preinstalled CAs -->
        <certificates src="system" />
        <!-- Additionaly trusted user added CAs -->
        <certificates src="user"/>
    </trust-anchors>
    </base-config>
~~~
(from [first answer](https://stackoverflow.com/a/72280826))

Not 100% sure that this is the cause, the same error would come for any exception but I think it is likely the cause.

You think this setting is something reasonable to add?

If not, don't worry, I can either build from source or use my public server with a real certificate.

**Comment by LiamMorrow** [comment]
Yeah I'd say that's super reasonable to add, thanks for letting me know!  I just realised you probably had to debug that without any sort of error logging, that would have been annoying.

I'll add some logs when it fails so hopefully logcat could help in future.

I'll do a release once this is merged:
https://github.com/LiamMorrow/LiftLog/pull/274

**Comment by bschwehn** [comment]
Thank you, it works well now for me!

**Comment by mBet608** [comment]
Are there any plans to support iCloud? I’m less concerned about backups and more interested in seamless syncing between my iPhone, iPad, and Mac.
